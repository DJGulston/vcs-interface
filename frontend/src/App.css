/*
  Page header is fixed, allowing the heading to stay in place while the user scrolls down.
  Z index of the header is set to 100 to ensure that it is displayed above all other
  elements on the page.
*/
.main-header {
  position: fixed;
  width: 100%;
  top:0;
  left:0;
  z-index: 100;
  background-color: rgb(50, 50, 50);
  border-bottom: 1px solid rgb(100, 100, 100);
}

h1 {
  color: lightcoral;
}

.header-item-section {
  position: relative;
  height: 90px;
}

/* Website name placed on the left hand side of the header. */
.header-item-1 {
  margin: 0;
  position: absolute;
  top: 50%;
  left: 20%;
  -ms-transform: translate(-80%, -50%);
  transform: translate(-80%, -50%);
  color:lightcoral;
  font-size: 30px;
  font-weight: bold;
}

/* Navigation menu placed on the right hand side of the header. */
.header-item-2 {
  margin: 0;
  position: absolute;
  top: 50%;
  left: 85%;
  -ms-transform: translate(-10%, -50%);
  transform: translate(-10%, -50%);
}

/* Navigation Menu link item is lightcoral. */
.link-item {
  color:lightcoral;
  font-size: 25px;
  font-weight: bold;
  text-decoration: none;
}

.form-block {
  display:block;
}

/* Displays textfield label within the 20% of the block space. */
.form-label {
  width:20%;
  display: inline-block;
}

/* Displays textfield input within the remaining 80% of the block space. */
.form-input {
  width:80%;
  display: inline-block;
}

/* Displays a red error alert. */
.input-alert {
  text-align: center;
  color:red;
  border: 1px red solid;
  border-radius: 6px;
  background-color: pink;
  padding:4px;
  margin-bottom: 6px;
  display:inline-flex;
}

.search-block {
  width: 50%;
  margin-left: auto;
  margin-right: auto;
}

.page-heading {
  text-align: center;
}

/* Close X symbol in the error alerts set to a slightly darker red. */
.close-x {
  margin-left: 5px;
  margin-right: 2px;
  color:rgb(181, 0, 0);
}

/* Displays list of repos in 3-column grid format. */
.repo-list {
  margin:20px;
  display: grid;
  grid-template-columns: 1fr 1fr 1fr;
  gap:30px;
}

.repo-section {
  border:1px solid whitesmoke;
  background-color: rgb(50, 50, 50);
  padding:25px;
  border-radius: 20px;
}

/* All links that take us to webpages outside of our website are limegreen. */
.external-link {
  color: limegreen;
}

/* Displays list of users in a 2-column grid format. */
.user-list {
  margin: 30px;
  display: grid;
  grid-template-columns: 1fr 1fr;
  gap: 30px;
}

.user-section {
  border: 1px solid whitesmoke;
  background-color: rgb(50, 50, 50);
  border-radius: 20px;
  padding: 30px;
  text-align: center;
}

/* All links that navigate us between different pages of our website are skyblue. */
.internal-link {
  text-decoration: none;
  color:skyblue;
  font-weight: bold;
}

/* Creates circular profile pictures for the GitHub and GitLab profile. */
.avatar-img {
  width: 120px;
  border-radius: 60px;
  margin:20px auto;
  margin-top: 0px;
  display: block;
}

.commit-message {
  margin: 25px;
}

/* Underline all commit message list items. */
.commit-message > li {
  text-decoration: underline;
}

/* Bullet point for the each list item in the commit message list is set to bold. */
.commit-message > li::marker {
  font-weight: bold;
}

.secondary-heading {
  margin-top: 30px;
}

.commit-message-label {
  color:lightgreen;
  font-weight: bold;
}

/* Creates a circular loading icon out of colored borders. */
.loading-icon {
  display: block;
  width:40px;
  height:40px;
  border:10px solid lightgrey;
  margin:40px auto;
  border-radius:20px;
  border-top: 10px solid blue;
  animation:icon-spin 1.5s linear infinite;
}

/* Animation for the loading icon. */
@keyframes icon-spin {
  0% { transform:rotate(0deg); }

  100% { transform:rotate(360deg); }
}