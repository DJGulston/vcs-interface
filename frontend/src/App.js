import './App.css';
import React, {Component} from 'react';
import { Route, Routes, BrowserRouter } from 'react-router-dom';
import {NavMenu} from './components/NavMenu.js';
import {UserSearch} from './components/UserSearch.js';
import {GitHubRepos} from './components/GitHubRepos.js';
import {GitLabRepos} from './components/GitLabRepos.js';
import {GitHubCommits} from './components/GitHubCommits.js';
import {GitLabCommits} from './components/GitLabCommits.js';
import {Container, Row, Col} from 'react-bootstrap';

class App extends Component {

  constructor(props) {

    super(props);

    this.state = {

      // Stores GitHub and GitLab user information.
      gitHubUserResult: props.gitHubUserResult,
      gitLabUserResult: props.gitLabUserResult,
      
      // Determines whether the GitHub and GitLab user information has been retrieved.
      gitHubUserRetrieved: props.gitHubUserRetrieved,
      gitLabUserRetrieved: props.gitLabUserRetrieved,
      
      // Stores the GitHub and GitLab repository information for a particular user.
      gitHubReposResult: props.gitHubReposResult,
      gitLabReposResult: props.gitLabReposResult,
      
      // Determines whether a user has typed in a username or not.
      searchHasInput: props.searchHasInput,

      // Stores the searched username.
      username: props.username,

      // Determines whether the GitHub and GitLab search results have loaded already.
      gitHubLoading: props.gitHubLoading,
      gitLabLoading: props.gitLabLoading,

      // Determines whether a username exists or not in GitHub and GitLab.
      gitHubUserFound: props.gitHubUserFound,
      gitLabUserFound: props.gitLabUserFound,

      // Determines whether the GitHub and GitLab repo information has loaded already.
      gitHubReposLoading: props.gitHubReposLoading,
      gitLabReposLoading: props.gitLabReposLoading,

      // Stores the selected GitHub and GitLab repo names.
      selectedGitHubRepoName: props.selectedGitHubRepoName,
      selectedGitLabRepoName: props.selectedGitLabRepoName,

      // Stores the GitHub and GitLab commits results for a particular selected repo name.
      gitHubCommitsResult: props.gitHubCommitsResult,
      gitLabCommitsResult: props.gitLabCommitsResult,

      // Determines whether the GitHub and GitLab commits results have loaded already.
      gitHubCommitsLoading: props.gitHubCommitsLoading,
      gitLabCommitsLoading: props.gitLabCommitsLoading,

      // Determines whether a JSON response could be retrieve from the GitHub and GitLab servers.
      isGitHubJsonResponse: props.isGitHubJsonResponse,
      isGitLabJsonResponse: props.isGitLabJsonResponse

    }

    // Functions that set the GitHub and GitLab JSON response booleans states.
    this.setIsGitHubJsonResponse = this.setIsGitHubJsonResponse.bind(this);
    this.setIsGitLabJsonResponse = this.setIsGitLabJsonResponse.bind(this);

    // Functions that set the JSON results for the GitHub and GitLab user details.
    this.setGitHubUserResult = this.setGitHubUserResult.bind(this);
    this.setGitLabUserResult = this.setGitLabUserResult.bind(this);

    // Functions that set JSON result for the GitHub and GitLab repository details.
    this.setGitHubReposResult = this.setGitHubReposResult.bind(this);
    this.setGitLabReposResult = this.setGitLabReposResult.bind(this);

    // Function that sets a boolean that determines whether a username entered input
    // into the search bar on the Search homepage.
    this.setSearchHasInput = this.setSearchHasInput.bind(this);

    // Functions that sets the searched username.
    this.setUsername = this.setUsername.bind(this);

    // Functions that set booleans that determine whether the GitHub and GitLab
    // user information has been retrieved.
    this.setGitHubUserRetrieved = this.setGitHubUserRetrieved.bind(this);
    this.setGitLabUserRetrieved = this.setGitLabUserRetrieved.bind(this);

    // Functions that set booleans that determine whether the GitHub and GitLab Search
    // results are still loading (being retrieved from the API).
    this.setGitHubLoading = this.setGitHubLoading.bind(this);
    this.setGitLabLoading = this.setGitLabLoading.bind(this);

    // Functions that set booleans that determine whether the searched username exists
    // on GitHub and GitLab.
    this.setGitHubUserFound = this.setGitHubUserFound.bind(this);
    this.setGitLabUserFound = this.setGitLabUserFound.bind(this);

    // Functions that set booleans that determine whether the GitHub and GitLab repo
    // results are still loading (being retrieved from the API).
    this.setGitHubReposLoading = this.setGitHubReposLoading.bind(this);
    this.setGitLabReposLoading = this.setGitLabReposLoading.bind(this);

    // Functions that set the selected GitHub and GitLab repository names.
    this.setSelectedGitHubRepoName = this.setSelectedGitHubRepoName.bind(this);
    this.setSelectedGitLabRepoName = this.setSelectedGitLabRepoName.bind(this);

    // Functions that set the GitHub and GitLab commits results for a selected
    // repository.
    this.setGitHubCommitsResult = this.setGitHubCommitsResult.bind(this);
    this.setGitLabCommitsResult = this.setGitLabCommitsResult.bind(this);

    // Functions that set the booleans that determine whether the GitHub and GitLab commits
    // results are still loading (being retrieved from the API).
    this.setGitHubCommitsLoading = this.setGitHubCommitsLoading.bind(this);
    this.setGitLabCommitsLoading = this.setGitLabCommitsLoading.bind(this);

  }

  // Group of functions below that set the states.

  setIsGitHubJsonResponse(isGitHubJsonResponse) {
    this.setState({
      isGitHubJsonResponse: isGitHubJsonResponse
    });
  }

  setIsGitLabJsonResponse(isGitLabJsonResponse) {
    this.setState({
      isGitLabJsonResponse: isGitLabJsonResponse
    });
  }

  setGitHubCommitsLoading(gitHubCommitsLoading) {
    this.setState({
      gitHubCommitsLoading: gitHubCommitsLoading
    });
  }

  setGitLabCommitsLoading(gitLabCommitsLoading) {
    this.setState({
      gitLabCommitsLoading: gitLabCommitsLoading
    });
  }

  setGitHubCommitsResult(gitHubCommitsResult) {
    this.setState({
      gitHubCommitsResult: gitHubCommitsResult
    });
  }

  setGitLabCommitsResult(gitLabCommitsResult) {
    this.setState({
      gitLabCommitsResult: gitLabCommitsResult
    });
  }

  setSelectedGitHubRepoName(selectedGitHubRepoName) {
    this.setState({
      selectedGitHubRepoName: selectedGitHubRepoName
    });
  }

  setSelectedGitLabRepoName(selectedGitLabRepoName) {
    this.setState({
      selectedGitLabRepoName: selectedGitLabRepoName
    });
  }

  setGitLabReposLoading(gitLabReposLoading) {
    this.setState({
      gitLabReposLoading: gitLabReposLoading
    });
  }

  setGitHubReposLoading(gitHubReposLoading) {
    this.setState({
      gitHubReposLoading: gitHubReposLoading
    });
  }

  setGitHubUserFound(gitHubUserFound) {
    this.setState({
      gitHubUserFound: gitHubUserFound
    });
  }

  setGitLabUserFound(gitLabUserFound) {
    this.setState({
      gitLabUserFound: gitLabUserFound
    });
  }

  setGitHubLoading(gitHubLoading) {
    this.setState({
      gitHubLoading: gitHubLoading
    });
  }

  setGitLabLoading(gitLabLoading) {
    this.setState({
      gitLabLoading: gitLabLoading
    });
  }

  setGitHubUserResult(gitHubUserResult) {
    this.setState({
      gitHubUserResult: gitHubUserResult
    });
  }

  setGitHubReposResult(gitHubReposResult) {
    this.setState({
      gitHubReposResult: gitHubReposResult
    });
  }

  setGitLabUserResult(gitLabUserResult) {
    this.setState({
      gitLabUserResult: gitLabUserResult
    });
  }

  setSearchHasInput(searchHasInput) {
    this.setState({
      searchHasInput: searchHasInput
    });
  }

  setUsername(username) {
    this.setState({
      username: username
    });
  }

  setGitHubUserRetrieved(gitHubUserRetrieved) {
    this.setState({
      gitHubUserRetrieved: gitHubUserRetrieved
    });
  }

  setGitLabReposResult(gitLabReposResult) {
    this.setState({
      gitLabReposResult: gitLabReposResult
    });
  }

  setGitLabUserRetrieved(gitLabUserRetrieved) {
    this.setState({
      gitLabUserRetrieved: gitLabUserRetrieved
    });
  }

  /**
   * Renders the VCS Interface page.
   * @returns 
   */
  render() {
    return (
      <BrowserRouter>
        <div>
    
          {/*
            Displays the header at the top of the page. Header includes a navigation menu that allows the user
            to navigate back to the Search homepage.
          */}
          <header className='main-header'>

            <div className='header-item-section'>
              <span className='header-item-1'>VCS Interface</span>
              <NavMenu />
            </div>

          </header>

          {/* Container that displays the main content of the page, i.e. all the routes of the website. */}
          <Container style={{
            marginTop:'90px'
          }}>
            <Row>
              <Col md={12}>
                <Routes>

                  {/* Route of the home page that allows the user to search GitHub and GitLab for a particular username. */}
                  <Route exact path='/' element={<UserSearch setGitHubUserResult={this.setGitHubUserResult} 
                                                            setGitHubReposResult={this.setGitHubReposResult} 
                                                            setGitLabUserResult={this.setGitLabUserResult}
                                                            setGitLabReposResult={this.setGitLabReposResult} 
                                                            setSearchHasInput={this.setSearchHasInput}
                                                            setUsername={this.setUsername}
                                                            setGitHubUserRetrieved={this.setGitHubUserRetrieved}
                                                            setGitLabUserRetrieved={this.setGitLabUserRetrieved}
                                                            setGitHubLoading={this.setGitHubLoading}
                                                            setGitLabLoading={this.setGitLabLoading}
                                                            setGitHubUserFound={this.setGitHubUserFound}
                                                            setGitLabUserFound={this.setGitLabUserFound}
                                                            setGitHubReposLoading={this.setGitHubReposLoading}
                                                            setGitLabReposLoading={this.setGitLabReposLoading}
                                                            setIsGitHubJsonResponse={this.setIsGitHubJsonResponse}
                                                            setIsGitLabJsonResponse={this.setIsGitLabJsonResponse}
                                                            gitHubUserResult={this.state.gitHubUserResult}
                                                            gitLabUserResult={this.state.gitLabUserResult}
                                                            searchHasInput={this.state.searchHasInput}
                                                            username={this.state.username}
                                                            gitHubUserRetrieved={this.state.gitHubUserRetrieved}
                                                            gitLabUserRetrieved={this.state.gitLabUserRetrieved}
                                                            gitHubLoading={this.state.gitHubLoading}
                                                            gitLabLoading={this.state.gitLabLoading}
                                                            gitHubUserFound={this.state.gitHubUserFound}
                                                            gitLabUserFound={this.state.gitLabUserFound}
                                                            isGitHubJsonResponse={this.state.isGitHubJsonResponse}
                                                            isGitLabJsonResponse={this.state.isGitLabJsonResponse} />} />

                  {/* Route for the GitHub repository page that displays the list of repositories for the searched user. */}
                  <Route exact path='/repos/github' element={<GitHubRepos gitHubReposResult={this.state.gitHubReposResult}
                                                                          gitHubReposLoading={this.state.gitHubReposLoading}
                                                                          setGitHubCommitsLoading={this.setGitHubCommitsLoading}
                                                                          setGitHubCommitsResult={this.setGitHubCommitsResult}
                                                                          setSelectedGitHubRepoName={this.setSelectedGitHubRepoName}
                                                                          username={this.state.username} />} />

                  {/* Route for the GitLab repository page that displays the list of repositories for the searched user. */}
                  <Route exact path='/repos/gitlab' element={<GitLabRepos gitLabReposResult={this.state.gitLabReposResult}
                                                                          gitLabReposLoading={this.state.gitLabReposLoading}
                                                                          setGitLabCommitsLoading={this.setGitLabCommitsLoading}
                                                                          setGitLabCommitsResult={this.setGitLabCommitsResult}
                                                                          setSelectedGitLabRepoName={this.setSelectedGitLabRepoName} />} />

                  {/* Route for the GitHub commits page that displays the list of the 5 most recent commits for the searched user. */}
                  <Route exact path='/commits/github' element={<GitHubCommits selectedGitHubRepoName={this.state.selectedGitHubRepoName}
                                                                              gitHubCommitsResult={this.state.gitHubCommitsResult}
                                                                              gitHubCommitsLoading={this.state.gitHubCommitsLoading} />} />

                  {/* Route for the GitLab commits page that displays the list of the 5 most recent commits for the searched user. */}
                  <Route exact path='/commits/gitlab' element={<GitLabCommits selectedGitLabRepoName={this.state.selectedGitLabRepoName}
                                                                              gitLabCommitsResult={this.state.gitLabCommitsResult}
                                                                              gitLabCommitsLoading={this.state.gitLabCommitsLoading} />} />

                </Routes>
              </Col>
            </Row>
          </Container>
          

        </div>
      </BrowserRouter>
    );
  }
  
}

export default App;

/*

References:

How to use useState in React:
- https://reactjs.org/docs/hooks-state.html

How to make a loading icon in CSS:
- https://www.w3schools.com/howto/howto_css_loader.asp

What is the base URL endpoint for GitLab API:
- https://stackoverflow.com/questions/39751840/cant-find-out-base-gitlab-api-base-url

How to get list of commits for repository in GitLab:
- https://docs.gitlab.com/ee/api/commits.html

How to get list of repositories in GitLab:
- https://docs.gitlab.com/ee/api/projects.html

How to get user info in GitLab:
- https://docs.gitlab.com/ee/api/users.html

How to get user info in GitHub:
- https://docs.github.com/en/rest/users/users?apiVersion=2022-11-28#get-a-user

How to get list of repositories in GitHub:
- https://docs.github.com/en/rest/repos/repos?apiVersion=2022-11-28

How to get list of commits for repository in GitHub:
- https://docs.github.com/en/rest/commits/commits?apiVersion=2022-11-28

How to render an X unicode symbol in React:
- https://medium.com/@jamesncox/rendering-unicode-symbol-in-react-76b1b4ffcee4

How to display arrows unicode symbols in React:
- https://www.w3schools.com/charsets/ref_utf_arrows.asp

How to pass custom data through an HTML element using a custom data attribute:
- https://www.w3schools.com/tags/att_global_data.asp
- https://stackoverflow.com/questions/432174/how-to-store-arbitrary-data-for-some-html-tags

How to make numbers of ordered list bold:
- https://stackoverflow.com/questions/21369843/is-there-a-way-to-make-numbers-in-an-ordered-list-bold

How to check if a variable is an array:
- https://www.w3schools.com/jsref/jsref_isarray.asp

How to use grid display layout in CSS:
- https://www.w3schools.com/css/css_grid_container.asp

*/
