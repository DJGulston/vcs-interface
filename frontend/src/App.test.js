import App from './App';
import React from 'react';
import renderer from 'react-test-renderer';

/**
 * Snapshot test to see if the App component renders correctly.
 */
test('App component renders correctly', () => {
  const tree = renderer
  .create(<App gitHubUserResult={[]}
    gitLabUserResult={[]}
    gitHubUserRetrieved={false}
    gitLabUserRetrieved={false}
    gitHubReposResult={[]}
    gitLabReposResult={[]}
    searchHasInput={true}
    username={""}
    gitHubLoading={false}
    gitLabLoading={false}
    gitHubUserFound={true}
    gitLabUserFound={true}
    gitHubReposLoading={false}
    gitLabReposLoading={false}
    selectedGitHubRepoName={""}
    selectedGitLabRepoName={""}
    gitHubCommitsResult={[]}
    gitLabCommitsResult={[]}
    gitHubCommitsLoading={false}
    gitLabCommitsLoading={false}
    isGitHubJsonResponse={true}
    isGitLabJsonResponse={true} />)
  .toJSON();

  expect(tree).toMatchSnapshot();
});