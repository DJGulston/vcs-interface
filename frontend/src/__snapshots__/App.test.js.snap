// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component renders correctly 1`] = `
<div>
  <header
    className="main-header"
  >
    <div
      className="header-item-section"
    >
      <span
        className="header-item-1"
      >
        VCS Interface
      </span>
      <nav
        className="header-item-2"
      >
        <a
          className="link-item"
          href="/"
          onClick={[Function]}
        >
          Search
        </a>
      </nav>
    </div>
  </header>
  <div
    className="container"
    style={
      Object {
        "marginTop": "90px",
      }
    }
  >
    <div
      className="row"
    >
      <div
        className="col-md-12"
      >
        <div
          className="card"
          style={
            Object {
              "backgroundColor": "rgb(50, 50, 50)",
              "border": "1px solid rgb(100, 100, 100)",
              "borderRadius": "25px",
              "margin": "40px auto",
              "padding": "30px",
              "width": "75%",
            }
          }
        >
          <h1
            className="page-heading"
          >
            Search
          </h1>
          <div
            className="search-block"
          >
            <div
              className="form-block"
            >
              <label
                className="form-label"
                htmlFor="username"
              >
                Username
              </label>
              <input
                className="form-input"
                id="username"
                name="username"
                placeholder="Search for username..."
                type="text"
              />
            </div>
            <div
              className="form-block"
            >
              <div
                className="form-label"
              />
              <span
                className="form-input"
              >
                <button
                  onClick={[Function]}
                >
                  Search
                </button>
              </span>
            </div>
          </div>
          <div
            className="user-list"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
